objects/capi.go

```diff
@@ -0,0 +1,149 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package objects
+
+import (
+	"archive/tar"
+	"bufio"
+	"bytes"
+	"compress/gzip"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/kubernetes/pkg/kubectl/scheme"
+)
+
+// getCRDs should actually use kustomize to correctly build the manager yaml.
+// HACK: this is a hacked function
+func getCAPIYAML(version, capiImage string) (io.Reader, error) {
+	crds := []string{"crds", "rbac", "manager"}
+	releaseCode := fmt.Sprintf("https://github.com/kubernetes-sigs/cluster-api/archive/%s.tar.gz", version)
+
+	resp, err := http.Get(releaseCode)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+
+	gz, err := gzip.NewReader(resp.Body)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+
+	tgz := tar.NewReader(gz)
+	var buf bytes.Buffer
+
+	for {
```

Comment:
Can we run `tar -xvf` using `cmd.exec` instead?  That, I think, will greatly simplify this code.

Reply:
This was just moved, I created it in a previous PR. I'm not execing for the same reason @liztio is removing the exec to kubectl. It is easier to manage dependencies through go.mod. In both cases it is very possible. If we were to exec we'd have to first make sure the right version of kubectl / tar /whatever are installed. It adds unnecessary complication for the end user when we can do it in Go and be done with it.

Reply:
That's a valid reason that applies more to `kubectl` than utils like `tar` or any other linux utils we may want to use.

Reply:
we'd still need to walk the entire directory and read these files in. And we don't want to assume `tar` exists with the same arguments on every system we need to interact with. `cmd.Exec()` should always be a last resort IMO

Reply:
I think this is avoidable complexity. But, I am OK with doing this 😄